oasislmf/pytools/gul/io.py

```diff
@@ -2,283 +2,351 @@
+
+    """
+    group_id_rng_index = Dict.empty(nb_int32, nb_int64)
+    rec_idx_ptr = List([0])
+
+    return group_id_rng_index, rec_idx_ptr
+
+
+def read_getmodel_stream(stream_in, item_map, coverages, compute, seeds, buff_size=GETMODEL_STREAM_BUFF_SIZE):
+    """Read the getmodel output stream yielding data event by event.
 
     Args:
-        run_dir (str): path to the run directory.
         stream_in (buffer-like): input stream, e.g. `sys.stdin.buffer`.
-        buff_size (int): size in bytes of the read buffer. Default is 65536.
+        item_map (Dict[ITEM_MAP_KEY_TYPE, ITEM_MAP_VALUE_TYPE]): dict storing
+          the mapping between areaperil_id, vulnerability_id to item.
+        coverages (numpy.ndarray[coverage_type]): array with coverage data.
+        compute (numpy.array[int]): list of coverages to be computed.
+        seeds (numpy.array[int]): the random seeds for each coverage_id.
+        buff_size (int): size in bytes of the read buffer (see note). Default is GETMODEL_STREAM_BUFF_SIZE.
 
     Raises:
         ValueError: If the stream type is not 1.
 
     Yields:
-        int32,  numpy.array[damagecdf], numpy.array[int], numpy.array[ProbMean]:
-          event_id, array of damagecdf entries (areaperil_id, vulnerability_id) for this event,
-          number of bins in all cdfs of event_id, all the cdfs used in event_id.
+        int, int, numpy.array[items_data_type], numpy.array[oasis_float], numpy.array[int], int:
+          event_id, index of the last coverage_id stored in compute, item-related data,
+          cdf records, array with the indices of `rec` where each cdf record starts,
+          number of unique random seeds computed so far.
+
+    Note:
+        It is advisable to set buff_size as 2x the maximum pipe limit (65536 bytes)
+        to ensure that the stream is always read in the biggest possible chunks,
+        which nominally is the largest between the pipe limit and the remaining memory
+        to fill the memoryview.
+
     """
     # determine stream type
     stream_type = np.frombuffer(stream_in.read(4), dtype='i4')
 
-    # TODO: make sure the bit1 and bit 2-4 compliance is checked
     # see https://github.com/OasisLMF/ktools/blob/master/docs/md/CoreComponents.md
     if stream_type[0] != 1:
         raise ValueError(f"FATAL: Invalid stream type: expect 1, got {stream_type[0]}.")
 
-    # get damage bins from fileFIXME
-    static_path = os.path.join(run_dir, 'static')
-    damage_bins = get_damage_bins(static_path=static_path)
-
-    # maximum number of damage bins (individual items can have up to `total_bins` bins)
-    if damage_bins.shape[0] == 0:
-        max_Nbins = 1000
-    else:
-        max_Nbins = damage_bins.shape[0]
-
     # maximum number of entries is buff_size divided by the minimum entry size
     # (corresponding to a 1-bin only cdf)
-    min_size_cdf_entry = damagecdfrec_stream.size + ProbMean.size + 4
+    min_size_cdf_entry = damagecdfrec_stream.size + 4 + ProbMean.size
```

Comment:
I think this 4 is not correct if area_peril_id is an int64

Reply:
The 4 is for the int32 storing `Nbins`.
The areaperil_int is already accounted for in `damagecdfrec_stream`
```py
damagecdfrec_stream = nb.from_dtype(np.dtype([('event_id', np.int32),
                                              ('areaperil_id', areaperil_int),
                                              ('vulnerability_id', np.int32)
                                              ]))
```

So, I think this is correct. Please resolve if you agree.